lib/helpers/versions.js

```diff
@@ -74,7 +75,73 @@ function determineUpgrades(dep, currentVersion, config) {
     delete allUpgrades.pin;
   }
   // Return only the values - we don't need the keys anymore
-  return Object.keys(allUpgrades).map(key => allUpgrades[key]);
+  const pinnedUpgrades = Object.keys(allUpgrades).map(key => allUpgrades[key]);
+
+  // Return now if array is empty, or we can keep pinned version upgrades
+  if (pinnedUpgrades.length === 0 || config.pinVersions || !isRange(currentVersion)) {
+    return pinnedUpgrades;
+  }
+
+  // The user prefers to maintain ranges, so we need to unpin our upgrades
+  const semverParsed = semverUtils.parseRange(currentVersion);
+  if (semverParsed.length > 1) {
+    // We don't know how to support complex semver ranges, so don't upgrade
+    logger.warn(`Can't support upgrading complex range ${currentVersion}`);
+    return [];
+  }
+  // Start with empty set of range upgrades
+  const rangeUpgrades = [];
+  // We know we have a simple semver, now check which operator it is
+  const currentSemver = semverParsed[0];
+  // Loop through all upgrades and convert to ranges
+  pinnedUpgrades.forEach((pinnedUpgrade) => {
```

Comment:
I think I'll stick with the forEach logic. The two reasons for discarding an upgrade are fairly simply done.